### PR TITLE
SPCTR-3339 Improvement: Container - Added a fallback image option for the video background-type.

### PR DESCRIPTION
**Pull request by AbhijitNage123**
### Description

- https://brainstormforce.atlassian.net/browse/SPCTR-3339

### Screenshots

- https://d.pr/v/fbauaU

### Types of changes

- New feature (non-breaking change which adds functionality)

### How has this been tested?

- Drag and drop container.
- Go to Style > Background > Type > Video > Select video and fallback image
- When fallback image can replace the background video in case the video cannot be loaded.

### Checklist:
- [x] My code is tested
- [ ] My code follows accessibility standards 
- [ ] My code has proper inline documentation 
- [x] I've included any necessary tests 
- [ ] I've included developer documentation 
- [x] I've added proper labels to this pull request

### CODE REVIEW

File: includes/blocks/container/frontend.css.php [hunk 1]
```diff
@@ -131,8 +131,8 @@
 }
 
 $background_video_opacity_value = ( isset( $attr['overlayOpacity'] ) && 'none' !== $attr['overlayType'] && ( ( 'color' === $attr['overlayType'] && ! empty( $attr['backgroundVideoColor'] ) ) || ( 'gradient' === $attr['overlayType'] && ! empty( $attr['gradientValue'] ) ) ) ) ? 1 - $attr['overlayOpacity'] : 1;
-
-$selectors = array(
+$bg_video_image_fallback        = isset( $attr['backgroundVideoFallbackImage'] ) && isset( $attr['backgroundVideoFallbackImage']['url'] ) ? $attr['backgroundVideoFallbackImage']['url'] : '';
```
Comment (bsf-zanev) [inline includes/blocks/container/frontend.css.php:134]: Would this be better?

```suggestion
$bg_video_image_fallback        = ! empty( $attr['backgroundVideoFallbackImage']['url'] ) ? $attr['backgroundVideoFallbackImage']['url'] : '';
```

File: src/blocks/container/styling.js [hunk 2]
```diff
@@ -282,6 +283,7 @@ function styling( attributes, clientId, name, deviceType ) {
 		},
 		' .uagb-container__video-wrap video': {
 			'opacity': backgroundVideoOpacityValue,
+			'background': `url(${ backgroundVideoFallbackImage?.url }) 50 50;`,
```
Comment (bsf-zanev) [inline src/blocks/container/styling.js:286]: Let's render this only if `backgroundVideoFallbackImage?.url` exists

File: includes/blocks/container/frontend.css.php [hunk 3]
```diff
@@ -162,7 +162,8 @@
 		'fill' => UAGB_Helper::hex2rgba( $attr['bottomColor'], ( isset( $attr['bottomDividerOpacity'] ) && '' !== $attr['bottomDividerOpacity'] ) ? $attr['bottomDividerOpacity'] : 100 ),
 	),
 	'.uagb-block-' . $id . ' .uagb-container__video-wrap video' => array(
-		'opacity' => $background_video_opacity_value,
+		'opacity'    => $background_video_opacity_value,
+		'background' => 'url(' . $bg_video_image_fallback . ') 50% 50%;',
```
Comment (bsf-zanev) [inline includes/blocks/container/frontend.css.php:166]: Same here.

File: src/components/background/index.js [hunk 4]
```diff
@@ -88,10 +88,11 @@ const Background = ( props ) => {
 		yPositionOverlayMobile,
 		yPositionOverlayType,
 		yPositionOverlayTypeTablet,
-		yPositionOverlayTypeMobile,
+		yPositionOverlayTypeMobile,
```
Comment (bsf-zanev) [inline src/components/background/index.js:91]:
```suggestion
		yPositionOverlayTypeMobile,
```

File: src/components/background/index.js [hunk 5]
```diff
@@ -1492,6 +1510,20 @@ const Background = ( props ) => {
...
+				<UAGMediaPicker
+					slug={ 'image' }
+					allow={ [ 'image' ] }
+					onSelectImage={ onSelectBgFallbackImage }
+					backgroundImage={ backgroundVideoFallbackImage.value }
+					onRemoveImage={ onRemoveBgFallbackImage }
+					label={ __( 'Fallback Image', 'ultimate-addons-for-gutenberg' ) }
+					help={ __( 'This cover image will replace the background video in case that the video could not be loaded.', 'ultimate-addons-for-gutenberg' ) }
+				/>
+				{/* <ResponsiveUAGImage backgroundImage={backgroundVideoFallbackImage} setAttributes={setAttributes} /> */}
```
Comment (bsf-zanev) [inline src/components/background/index.js:1524]: Let's remove this commented code if it's not required.

**Review by bsf-zanev — COMMENTED** [review]
Added a few minor comments.

File: includes/blocks/container/frontend.css.php (hunk 1 above)
Comment (AbhijitNage123) [reply includes/blocks/container/frontend.css.php]: Resolved

File: src/components/background/index.js (hunk 4 above)
Comment (AbhijitNage123) [reply src/components/background/index.js]: Resolved

File: src/components/background/index.js (hunk 5 above)
Comment (AbhijitNage123) [reply src/components/background/index.js]: Resolved

File: src/blocks/container/styling.js (hunk 2 above)
Comment (AbhijitNage123) [reply src/blocks/container/styling.js]: Done

File: includes/blocks/container/frontend.css.php (hunk 3 above)
Comment (AbhijitNage123) [reply includes/blocks/container/frontend.css.php:169]: Done

File: src/blocks/container/styling.js [hunk 6]
```diff
@@ -280,11 +281,19 @@ function styling( attributes, clientId, name, deviceType ) {
 		' > .uagb-container__shape-bottom .uagb-container__shape-fill': {
 			'fill': bottomColor,
 		},
-		' .uagb-container__video-wrap video': {
-			'opacity': backgroundVideoOpacityValue,
-		},
```
Comment (bsf-zanev) [inline src/blocks/container/styling.js:285]: We can keep this here, since it's common for both cases.

File: src/blocks/container/styling.js [hunk 7]
```diff
@@ -280,11 +281,19 @@ function styling( attributes, clientId, name, deviceType ) {
...
+		selectors[' .uagb-container__video-wrap video'] = {
+			'opacity': backgroundVideoOpacityValue,
+			'background': `url(${backgroundVideoFallbackImage.url}) 50% 50%;`,
+		};
+	} else {
+		selectors[' .uagb-container__video-wrap video'] = {
+			'opacity': backgroundVideoOpacityValue,
+		};
+	}
+
```
Comment (bsf-zanev) [inline src/blocks/container/styling.js:296]:
```suggestion
	if ( backgroundVideoFallbackImage?.url ) {
		selectors[' .uagb-container__video-wrap video'] = {
			...selectors[' .uagb-container__video-wrap video'],
			'background': `url(${ backgroundVideoFallbackImage.url }) 50% 50%;`,
		};
	}

```

### OR

```suggestion
	if ( backgroundVideoFallbackImage?.url ){
		selectors[' .uagb-container__video-wrap video']['background'] = `url(${ backgroundVideoFallbackImage.url }) 50% 50%;`,
		};
	} 

```

File: includes/blocks/container/frontend.css.php [hunk 8]
```diff
@@ -161,11 +161,19 @@
 	'.uagb-block-' . $id . ' .uagb-container__shape.uagb-container__shape-bottom .uagb-container__shape-fill' => array(
 		'fill' => UAGB_Helper::hex2rgba( $attr['bottomColor'], ( isset( $attr['bottomDividerOpacity'] ) && '' !== $attr['bottomDividerOpacity'] ) ? $attr['bottomDividerOpacity'] : 100 ),
 	),
-	'.uagb-block-' . $id . ' .uagb-container__video-wrap video' => array(
-		'opacity' => $background_video_opacity_value,
-	),
```
Comment (bsf-zanev) [inline includes/blocks/container/frontend.css.php:166]: Same here.

**Review by bsf-zanev — COMMENTED** [review]
One minor concern, common in JS and PHP.

File: src/blocks/container/styling.js (hunk 6 above)
Comment (AbhijitNage123) [reply src/blocks/container/styling.js]: Done

File: src/blocks/container/styling.js (hunk 7 above)
Comment (AbhijitNage123) [reply src/blocks/container/styling.js]: Done working as expected

File: includes/blocks/container/frontend.css.php (hunk 8 above)
Comment (AbhijitNage123) [reply includes/blocks/container/frontend.css.php]: Done working as expected

**Review by bsf-zanev — APPROVED** [review]
Looks good to me.

**Review by adi3890 — APPROVED** [review]

**Pull request merged**